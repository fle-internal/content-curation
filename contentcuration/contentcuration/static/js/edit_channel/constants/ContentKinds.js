module.exports = [
  {
    kind: 'document',
  },
  {
    kind: 'topic',
  },
  {
    kind: 'video',
  },
  {
    kind: 'html5',
  },
  {
    kind: 'audio',
  },
  {
    kind: 'exercise',
  },
];
