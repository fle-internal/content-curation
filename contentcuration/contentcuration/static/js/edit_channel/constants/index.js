const ContentKinds = require('./ContentKinds');
const FormatPresets = require('./FormatPresets');
const Languages = require('./Languages');
const Licenses = require('./Licenses');
const MasteryModels = require('./MasteryModels');
const Roles = require('./Roles');

module.exports = {
  ContentKinds,
  FormatPresets,
  Languages,
  Licenses,
  MasteryModels,
  Roles,
};
