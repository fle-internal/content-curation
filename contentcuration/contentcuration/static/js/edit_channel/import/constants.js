export const PageTypes = {
  TREE_VIEW: 'TREE_VIEW',
  IMPORT_PREVIEW: 'IMPORT_PREVIEW',
  SEARCH_RESULTS: 'SEARCH_RESULTS',
};
