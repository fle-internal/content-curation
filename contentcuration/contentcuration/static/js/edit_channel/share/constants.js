export const Permissions = {
  OWNER: 'manage',
  EDIT: 'edit',
  VIEW_ONLY: 'view'
};

export const PermissionRanks = [
	{
		"shareMode": Permissions.VIEW_ONLY,
		"rank": 0,
		"field": 'viewers'
	},
	{
		"shareMode": Permissions.EDIT,
		"rank": 1,
		"field": 'editors'
	},
	{
		"shareMode": Permissions.OWNER,
		"rank": 2,
		"field": null
	},
]
