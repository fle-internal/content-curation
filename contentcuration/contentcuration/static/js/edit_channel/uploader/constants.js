export const TabNames = {
  DETAILS: 'DETAILS',
  PREVIEW: 'PREVIEW',
  QUESTIONS: 'QUESTIONS',
  PREREQUISITES: 'PREREQUISITES',
};

export const modes = {
  EDIT: 'EDIT',
  NEW_TOPIC: 'NEW_TOPIC',
  NEW_EXERCISE: 'NEW_EXERCISE',
  UPLOAD: 'UPLOAD',
  VIEW_ONLY: 'VIEW_ONLY',
};

export const AssessmentItemToolbarActions = {
  MOVE_ITEM_UP: 'MOVE_ITEM_UP',
  MOVE_ITEM_DOWN: 'MOVE_ITEM_DOWN',
  EDIT_ITEM: 'EDIT_ITEM',
  ADD_ITEM_ABOVE: 'ADD_ITEM_ABOVE',
  ADD_ITEM_BELOW: 'ADD_ITEM_BELOW',
  DELETE_ITEM: 'DELETE_ITEM',
};

export const AssessmentItemValidationErrors = {
  BLANK_QUESTION: 'BLANK_QUESTION',
  INVALID_NUMBER_OF_CORRECT_ANSWERS: 'INVALID_NUMBER_OF_CORRECT_ANSWERS',
};

// should correspond to backend types
export const AssessmentItemTypes = {
  SINGLE_SELECTION: 'single_selection',
  MULTIPLE_SELECTION: 'multiple_selection',
  TRUE_FALSE: 'true_false',
  INPUT_QUESTION: 'input_question',
};

export const AssessmentItemTypeLabels = {
  [AssessmentItemTypes.SINGLE_SELECTION]: 'Single selection',
  [AssessmentItemTypes.MULTIPLE_SELECTION]: 'Multiple selection',
  [AssessmentItemTypes.TRUE_FALSE]: 'True/False',
  [AssessmentItemTypes.INPUT_QUESTION]: 'Input question',
};
