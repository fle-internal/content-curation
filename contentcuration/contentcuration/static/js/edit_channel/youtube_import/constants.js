export const PageTypes = {
  SUBMIT_URL: 'SUBMIT_URL',
  CONFIRM: 'CONFIRM',
};

export const ImportStatus = {
  IDLE: 'IDLE',
  EXTRACTING: 'EXTRACTING',
  EXTRACT_ERROR: 'EXTRACT_ERROR',
  DOWNLOADING: 'DOWNLOADING',
  DOWNLOAD_ERROR: 'DOWNLOAD_ERROR',
  START_IMPORT: 'START_IMPORT',
  DONE: 'DONE',
};
