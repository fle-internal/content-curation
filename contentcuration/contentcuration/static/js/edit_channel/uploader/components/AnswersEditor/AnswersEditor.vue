<template>
  <div>
    <div class="grey--text text--darken-1 mb-3">
      Answers
    </div>

    <div v-if="!answers || !answers.length">
      No answers yet
    </div>

    <VCard
      v-for="(answer, answerIdx) in answers"
      :key="answerIdx"
      :style="getAnswerPointerStyle(answerIdx)"
      data-test="answer"
      @click="onAnswerClick($event, answerIdx)"
    >
      <div :class="getIndicatorClasses(answer)"></div>

      <VCardText>
        <VLayout>
          <VFlex xs8>
            <template v-if="!isAnswerOpen(answerIdx)">
              <!--
                VRadio cannot be used without VRadioGroup like VCheckbox but it can
                be solved by wrapping each VRadio to VRadioGroup
                https://github.com/vuetifyjs/vuetify/issues/2345
              -->
              <VRadioGroup
                v-if="hasOneCorrectAnswer"
                v-model="correctAnswersIndices"
              >
                <VRadio
                  :value="answerIdx"
                  :label="answer.answer"
                  data-test="answerRadio"
                />
              </VRadioGroup>

              <VCheckbox
                v-if="isMultipleSelection"
                :key="answerIdx"
                v-model="correctAnswersIndices"
                :value="answerIdx"
                :label="answer.answer"
                data-test="answerCheckbox"
              />

              <div
                v-if="isInputQuestion"
                class="input-question"
              >
                {{ answer.answer }}
              </div>
            </template>

            <template v-if="isAnswerOpen(answerIdx)">
              <VLayout>
                <VFlex xs1>
                  <VRadioGroup
                    v-if="hasOneCorrectAnswer"
                    v-model="correctAnswersIndices"
                  >
                    <VRadio
                      :value="answerIdx"
                      data-test="answerRadio"
                    />
                  </VRadioGroup>

                  <VCheckbox
                    v-if="isMultipleSelection"
                    :key="answerIdx"
                    v-model="correctAnswersIndices"
                    :value="answerIdx"
                    data-test="answerCheckbox"
                  />
                </VFlex>

                <VFlex xs10>
                  <VTextField
                    :value="answer.answer"
                    data-test="editAnswerTextInput"
                    @input="updateAnswerText($event, answerIdx)"
                  />
                </VFlex>
              </VLayout>
            </template>
          </VFlex>

          <VSpacer />
          <AssessmentItemToolbar
            :displayMenu="false"
            :displayEditIcon="false"
            :canMoveUp="!isAnswerFirst(answerIdx)"
            :canMoveDown="!isAnswerLast(answerIdx)"
            class="toolbar"
            @click="onToolbarClick($event, answerIdx)"
          />
        </VLayout>
      </VCardText>
    </VCard>

    <VBtn
      v-if="isEditingAllowed"
      flat
      data-test="newAnswerBtn"
      @click="addNewAnswer"
    >
      New answer
    </VBtn>
  </div>
</template>

<script>

  import { AssessmentItemTypes, AssessmentItemToolbarActions } from '../../constants';
  import {
    questionHasOneCorrectAnswer,
    getCorrectAnswersIndices,
    mapCorrectAnswers,
    swapElements,
    sanitizeAssessmentItemAnswers,
  } from '../../utils';
  import AssessmentItemToolbar from '../AssessmentItemToolbar/AssessmentItemToolbar.vue';

  export default {
    name: 'AnswersEditor',
    components: {
      AssessmentItemToolbar,
    },
    model: {
      prop: 'answers',
      event: 'update',
    },
    props: {
      questionKind: {
        type: String,
        validator: value => {
          return Object.values(AssessmentItemTypes).includes(value);
        },
        default: AssessmentItemTypes.SINGLE_SELECTION,
      },
      answers: {
        type: Array,
      },
      openAnswerIdx: {
        type: Number,
      },
    },
    data() {
      return {
        correctAnswersIndices: getCorrectAnswersIndices(this.questionKind, this.answers),
      };
    },
    computed: {
      isSingleSelection() {
        return this.questionKind === AssessmentItemTypes.SINGLE_SELECTION;
      },
      isMultipleSelection() {
        return this.questionKind === AssessmentItemTypes.MULTIPLE_SELECTION;
      },
      isTrueFalse() {
        return this.questionKind === AssessmentItemTypes.TRUE_FALSE;
      },
      isInputQuestion() {
        return this.questionKind === AssessmentItemTypes.INPUT_QUESTION;
      },
      hasOneCorrectAnswer() {
        return questionHasOneCorrectAnswer(this.questionKind);
      },
      isEditingAllowed() {
        return !this.isTrueFalse;
      },
    },
    watch: {
      answers() {
        this.correctAnswersIndices = getCorrectAnswersIndices(this.questionKind, this.answers);
      },
      correctAnswersIndices(newValue, oldValue) {
        if (oldValue !== null) {
          if (this.hasOneCorrectAnswer && newValue === oldValue) {
            return;
          }

          if (
            !this.hasOneCorrectAnswer &&
            JSON.stringify([...newValue].sort()) === JSON.stringify([...oldValue].sort())
          ) {
            return;
          }
        }

        const updatedAnswers = mapCorrectAnswers(this.answers, this.correctAnswersIndices);

        this.emitUpdate(updatedAnswers);
      },
    },
    methods: {
      isAnswerFirst(answerIdx) {
        return answerIdx === 0;
      },
      isAnswerLast(answerIdx) {
        return answerIdx === this.answers.length - 1;
      },
      isAnswerOpen(answerIdx) {
        return answerIdx === this.openAnswerIdx;
      },
      getAnswerPointerStyle(answerIdx) {
        return this.isEditingAllowed && !this.isAnswerOpen(answerIdx) ? { cursor: 'pointer' } : {};
      },
      getIndicatorClasses(answer) {
        const classes = ['indicator'];

        if (answer.correct) {
          classes.push('correct');
        } else {
          classes.push('wrong');
        }

        return classes;
      },
      emitOpen(answerIdx) {
        if (!this.isEditingAllowed) {
          return;
        }

        this.$emit('open', answerIdx);
      },
      emitClose() {
        this.$emit('close');
      },
      emitUpdate(updatedAnswers) {
        this.$emit('update', updatedAnswers);
      },
      moveAnswerUp(answerIdx) {
        if (this.isAnswerFirst(answerIdx)) {
          return;
        }

        let updatedAnswers = swapElements(this.answers, answerIdx, answerIdx - 1);
        updatedAnswers = updatedAnswers.map((answer, answerIdx) => {
          return {
            ...answer,
            order: answerIdx + 1,
          };
        });

        this.emitUpdate(updatedAnswers);
        if (this.isAnswerOpen(answerIdx)) {
          this.emitOpen(answerIdx - 1);
        } else if (this.isAnswerOpen(answerIdx - 1)) {
          this.emitOpen(answerIdx);
        }
      },
      moveAnswerDown(answerIdx) {
        if (this.isAnswerLast(answerIdx)) {
          return;
        }

        let updatedAnswers = swapElements(this.answers, answerIdx, answerIdx + 1);
        updatedAnswers = updatedAnswers.map((answer, answerIdx) => {
          return {
            ...answer,
            order: answerIdx + 1,
          };
        });

        this.emitUpdate(updatedAnswers);
        if (this.isAnswerOpen(answerIdx)) {
          this.emitOpen(answerIdx + 1);
        } else if (this.isAnswerOpen(answerIdx + 1)) {
          this.emitOpen(answerIdx);
        }
      },
      deleteAnswer(answerIdx) {
        let updatedAnswers = JSON.parse(JSON.stringify(this.answers));

        updatedAnswers.splice(answerIdx, 1);
        updatedAnswers = updatedAnswers.map((answer, answerIdx) => {
          return {
            ...answer,
            order: answerIdx + 1,
          };
        });

        this.emitUpdate(updatedAnswers);
        if (this.isAnswerOpen(answerIdx)) {
          this.emitClose();
        } else if (this.openAnswerIdx > answerIdx) {
          this.emitOpen(this.openAnswerIdx - 1);
        }
      },
      onAnswerClick(event, answerIdx) {
        // do not open answer on toolbar click
        if (event.target.closest('.toolbar') !== null) {
          return;
        }

        // do not open an answer on checkbox or radio click
        if (
          event.target &&
          (event.target.classList.contains('v-label') ||
            event.target.classList.contains('v-input--selection-controls__ripple'))
        ) {
          return;
        }

        this.emitOpen(answerIdx);
      },
      onToolbarClick(action, answerIdx) {
        switch (action) {
          case AssessmentItemToolbarActions.MOVE_ITEM_UP:
            this.moveAnswerUp(answerIdx);
            break;

          case AssessmentItemToolbarActions.MOVE_ITEM_DOWN:
            this.moveAnswerDown(answerIdx);
            break;

          case AssessmentItemToolbarActions.DELETE_ITEM:
            this.deleteAnswer(answerIdx);
            break;
        }
      },
      updateAnswerText(newAnswerText, answerIdx) {
        if (newAnswerText === this.answers[answerIdx].answer) {
          return;
        }

        const updatedAnswers = [...this.answers];
        updatedAnswers[answerIdx].answer = newAnswerText;

        this.emitUpdate(updatedAnswers);
      },
      addNewAnswer() {
        // primarily to disable adding more empty hints
        const updatedAnswers = sanitizeAssessmentItemAnswers(this.answers, true);

        const defaultCorrectState = this.isInputQuestion ? true : false;
        updatedAnswers.push({
          answer: '',
          correct: defaultCorrectState,
          order: updatedAnswers.length + 1,
        });

        this.emitUpdate(updatedAnswers);
        this.emitOpen(updatedAnswers.length - 1);
      },
    },
  };

</script>

<style lang="less" scoped>

  @import '../../../../../less/global-variables.less';

  .indicator {
    position: absolute;
    width: 4px;
    height: 100%;

    &.correct {
      background-color: @exercise-answer-correct;
    }

    &.wrong {
      background-color: @exercise-answer-wrong;
    }
  }

  .input-question {
    padding: 23.5px;
  }

</style>
