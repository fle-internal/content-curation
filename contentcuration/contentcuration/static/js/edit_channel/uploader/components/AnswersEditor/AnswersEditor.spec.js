import { shallowMount, mount } from '@vue/test-utils';

import { AssessmentItemTypes } from '../../constants';
import AnswersEditor from './AnswersEditor';

const clickNewAnswerBtn = wrapper => {
  wrapper
    .find('[data-test=newAnswerBtn]')
    .find('button')
    .trigger('click');
};

const rendersNewAnswerBtn = wrapper => {
  return wrapper.contains('[data-test=newAnswerBtn]');
};

const clickAnswer = (wrapper, answerIdx) => {
  wrapper
    .findAll('[data-test=answer]')
    .at(answerIdx)
    .trigger('click');
};

const clickMoveAnswerUp = (wrapper, answerIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconArrowUp]')
    .at(answerIdx)
    .trigger('click');
};

const clickMoveAnswerDown = (wrapper, answerIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconArrowDown]')
    .at(answerIdx)
    .trigger('click');
};

const clickDeleteAnswer = (wrapper, answerIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconDelete]')
    .at(answerIdx)
    .trigger('click');
};

const updateOpenAnswerText = (wrapper, newAnswerText) => {
  // only one answer can be open
  wrapper.find('[data-test=editAnswerTextInput]').setValue(newAnswerText);
};

describe('AnswersEditor', () => {
  let wrapper;

  it('smoke test', () => {
    const wrapper = shallowMount(AnswersEditor);

    expect(wrapper.isVueInstance()).toBe(true);
  });

  describe('with no answers', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          answers: [],
        },
      });
    });

    it('renders a placeholder', () => {
      expect(wrapper.html()).toContain('No answers yet');
    });
  });

  describe('for a single selection question', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });
    });

    it('renders answers as radio selects', () => {
      const inputs = wrapper.findAll('input');
      const labels = wrapper.findAll('label');

      expect(inputs.length).toBe(2);
      expect(labels.length).toBe(2);

      for (let n in [0, 1]) {
        expect(inputs.at(n).attributes()['type']).toBe('radio');
      }

      expect(labels.at(0).text()).toBe('Mayonnaise (I mean you can, but...)');
      expect(labels.at(1).text()).toBe('Peanut butter');
    });

    it('renders a correct answer as checked', () => {
      const inputs = wrapper.findAll('input');

      expect(inputs.at(0).element.checked).toBe(true);
      expect(inputs.at(1).element.checked).toBe(false);
    });

    it('renders new answer button', () => {
      expect(rendersNewAnswerBtn(wrapper)).toBe(true);
    });

    describe('on new answer button click', () => {
      beforeEach(() => {
        clickNewAnswerBtn(wrapper);
      });

      it('emits update event with a payload containing all answers + new answer which is wrong by default', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual([
          { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
          { answer: 'Peanut butter', correct: false, order: 2 },
          { answer: '', correct: false, order: 3 },
        ]);
      });
    });
  });

  describe('for a multiple selection question', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.MULTIPLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
            { answer: 'Jelly', correct: true, order: 3 },
          ],
        },
      });
    });

    it('renders answers as checkboxes', () => {
      const inputs = wrapper.findAll('input');
      const labels = wrapper.findAll('label');

      expect(inputs.length).toBe(3);
      expect(labels.length).toBe(3);

      for (let n in [0, 1, 2]) {
        expect(inputs.at(n).attributes()['type']).toBe('checkbox');
      }

      expect(labels.at(0).text()).toBe('Mayonnaise (I mean you can, but...)');
      expect(labels.at(1).text()).toBe('Peanut butter');
      expect(labels.at(2).text()).toBe('Jelly');
    });

    it('renders all correct answers as checked', () => {
      const inputs = wrapper.findAll('input');

      expect(inputs.at(0).element.checked).toBe(true);
      expect(inputs.at(1).element.checked).toBe(false);
      expect(inputs.at(2).element.checked).toBe(true);
    });

    it('renders new answer button', () => {
      expect(rendersNewAnswerBtn(wrapper)).toBe(true);
    });

    describe('on new answer button click', () => {
      beforeEach(() => {
        clickNewAnswerBtn(wrapper);
      });

      it('emits update event with a payload containing all answers + new answer which is wrong by default', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual([
          { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
          { answer: 'Peanut butter', correct: false, order: 2 },
          { answer: 'Jelly', correct: true, order: 3 },
          { answer: '', correct: false, order: 4 },
        ]);
      });
    });
  });

  describe('for a true/false question', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.TRUE_FALSE,
          answers: [
            { answer: 'True', correct: false, order: 1 },
            { answer: 'False', correct: true, order: 2 },
          ],
        },
      });
    });

    it('renders answers as radio selects', () => {
      const inputs = wrapper.findAll('input');
      const labels = wrapper.findAll('label');

      expect(inputs.length).toBe(2);
      expect(labels.length).toBe(2);

      for (let n in [0, 1]) {
        expect(inputs.at(n).attributes()['type']).toBe('radio');
      }

      expect(labels.at(0).text()).toBe('True');
      expect(labels.at(1).text()).toBe('False');
    });

    it('renders a correct answer as checked', () => {
      const inputs = wrapper.findAll('input');

      expect(inputs.at(0).element.checked).toBe(false);
      expect(inputs.at(1).element.checked).toBe(true);
    });

    it('does not render new answer button', () => {
      expect(rendersNewAnswerBtn(wrapper)).toBe(false);
    });
  });

  describe('for an input question', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.INPUT_QUESTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: true, order: 2 },
          ],
        },
      });
    });

    it('renders all possible answers', () => {
      expect(wrapper.html()).toContain('Mayonnaise (I mean you can, but...)');
      expect(wrapper.html()).toContain('Peanut butter');
    });

    it('renders new answer button', () => {
      expect(rendersNewAnswerBtn(wrapper)).toBe(true);
    });

    describe('on new answer button click', () => {
      beforeEach(() => {
        clickNewAnswerBtn(wrapper);
      });

      it('emits update event with a payload containing all answers + new answer which is correct', () => {
        expect(wrapper.emitted().update).toBeTruthy();
        expect(wrapper.emitted().update.length).toBe(1);
        expect(wrapper.emitted().update[0][0]).toEqual([
          { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
          { answer: 'Peanut butter', correct: true, order: 2 },
          { answer: '', correct: true, order: 3 },
        ]);
      });
    });
  });

  describe('on answer text update', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });

      clickAnswer(wrapper, 1);
      updateOpenAnswerText(wrapper, 'Irish butter');
    });

    it('emits update event with a payload containing updated answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
        { answer: 'Irish butter', correct: false, order: 2 },
      ]);
    });
  });

  describe('on correct answer change', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });

      clickAnswer(wrapper, 1);
      wrapper
        .findAll('[data-test=answerRadio]')
        .at(1)
        .setChecked();
    });

    it('emits update event with a payload containing updated answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { answer: 'Mayonnaise (I mean you can, but...)', correct: false, order: 1 },
        { answer: 'Peanut butter', correct: true, order: 2 },
      ]);
    });
  });

  describe('on move answer up click', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });

      clickMoveAnswerUp(wrapper, 1);
    });

    it('emits update event with a payload containing updated and properly ordered answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { answer: 'Peanut butter', correct: false, order: 1 },
        { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 2 },
      ]);
    });
  });

  describe('on move answer down click', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });

      clickMoveAnswerDown(wrapper, 0);
    });

    it('emits update event with a payload containing updated and properly ordered answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { answer: 'Peanut butter', correct: false, order: 1 },
        { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 2 },
      ]);
    });
  });

  describe('on delete answer click', () => {
    beforeEach(() => {
      wrapper = mount(AnswersEditor, {
        propsData: {
          questionKind: AssessmentItemTypes.SINGLE_SELECTION,
          answers: [
            { answer: 'Mayonnaise (I mean you can, but...)', correct: true, order: 1 },
            { answer: 'Peanut butter', correct: false, order: 2 },
          ],
        },
      });

      clickDeleteAnswer(wrapper, 0);
    });

    it('emits update event with a payload containing updated and properly ordered answers', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { answer: 'Peanut butter', correct: false, order: 1 },
      ]);
    });
  });
});
