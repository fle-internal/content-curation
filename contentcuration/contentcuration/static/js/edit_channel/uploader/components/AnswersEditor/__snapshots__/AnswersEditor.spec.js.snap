// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswersEditor for a multiple selection question toolbar contains move up/down and delete icons 1`] = `
<div data-test="toolbar" class="toolbar"><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
    <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_up</i></div>
  </button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div></span><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
  <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_down</i></div>
</button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div></span><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
  <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
</button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div></span>
<!---->
</div>
`;

exports[`AnswersEditor for a single selection question toolbar contains move up/down and delete icons 1`] = `
<div data-test="toolbar" class="toolbar"><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
    <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_up</i></div>
  </button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div></span><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
  <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_down</i></div>
</button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div></span><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
  <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
</button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div></span>
<!---->
</div>
`;

exports[`AnswersEditor for a true/false question renders empty toolbar 1`] = `
<div data-test="toolbar" class="toolbar">
  <!---->
</div>
`;

exports[`AnswersEditor for an input question toolbar contains only delete icon 1`] = `
<div data-test="toolbar" class="toolbar"><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
    <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
  </button><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div></span>
<!---->
</div>
`;
